### PR TITLE
Introduce queries pagination

### PR DESCRIPTION
**Pull request by nickaleks**
### Description of the Change
This Pull Request adds pagination to several queries. This allows to work on large data sets without returning all the data in a single request

### Benefits
Better interface for queries

### Possible Drawbacks 
Nonne

### CODE REVIEW

File: irohad/ametsuchi/impl/postgres_query_executor.cpp [hunk 1]
```diff
@@ -339,6 +340,115 @@ namespace iroha {
...
+            }
+            // If 0 transactions are returned, we assume that hash is invalid.
+            // Since query with valid hash is guaranteed to return at least one
+            // transaction
+            if (first_hash and response_txs.empty()) {
+              auto error = (boost::format("invalid pagination hash: %s")
+                            % first_hash->hex())
+                               .str();
+              // TODO: IR-82 nickaleks 7.12.18
+              // add status code for invalid pagination
```
Comment (Akvinikym) [inline irohad/ametsuchi/impl/postgres_query_executor.cpp]: Remove TODO please

File: shared_model/bindings/model_query_builder.cpp [hunk 2]
```diff
@@ -37,15 +37,20 @@ namespace shared_model {
     }
 
     ModelQueryBuilder ModelQueryBuilder::getAccountTransactions(
-        const interface::types::AccountIdType &account_id) {
-      return ModelQueryBuilder(builder_.getAccountTransactions(account_id));
+        const interface::types::AccountIdType &account_id,
+        interface::types::TransactionsNumberType page_size,
+        const boost::optional<interface::types::HashType> &first_hash) {
```
Comment (lebdron) [inline shared_model/bindings/model_query_builder.cpp]: It is highly likely that optional is not wrapped with SWIG in an expected way. Please check java bindings and comment if it was successful or not.

File: shared_model/interfaces/query_responses/impl/transactions_page_response.cpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+                         .append("all transactions size",
+                                 std::to_string(allTransactionsSize()));
+      if (nextTxHash()) {
+        return builder.append("next tx hash", nextTxHash()->hex()).finalize();
+      }
+      return builder.finalize();
+    }
+
+    bool TransactionsPageResponse::operator==(const ModelType &rhs) const {
+      return transactions() == rhs.transactions();
```
Comment (lebdron) [inline shared_model/interfaces/query_responses/impl/transactions_page_response.cpp]: Why is it not required to also compare nextTxHash and allTransactionsSize?

File: test/module/irohad/ametsuchi/postgres_query_executor_test.cpp [hunk 4]
```diff
@@ -29,23 +41,58 @@
...
+                .quorum(1)
+                .jsonData(R"({"id@domain": {"key": "value"}})")
+                .build());
+
+  template <typename T>
+  using ConstRef = const T &;
+
+  // old compilers fail to deduce that a variant contains reference
+  template <typename T, typename Variant>
+  ConstRef<T> getRef(Variant &&v) {
```
Comment (lebdron) [inline test/module/irohad/ametsuchi/postgres_query_executor_test.cpp]: Method is unused, please remove.

File: test/module/irohad/ametsuchi/postgres_query_executor_test.cpp [hunk 5]
```diff
@@ -29,23 +41,58 @@
...
+  const std::unique_ptr<Account> account2 =
+      clone(TestAccountBuilder()
+                .domainId(domain->domainId())
+                .accountId("id2@" + domain->domainId())
+                .quorum(1)
+                .jsonData(R"({"id@domain": {"key": "value"}})")
+                .build());
+
+  template <typename T>
+  using ConstRef = const T &;
```
Comment (lebdron) [inline test/module/irohad/ametsuchi/postgres_query_executor_test.cpp]: Same with unused, please remove.

**Pull request merged**